dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/interceptors/LoggedCallInterceptor.java

```diff
@@ -0,0 +1,106 @@
+package edu.harvard.iq.dataverse.interceptors;
+
+import edu.harvard.iq.dataverse.DataverseRequestServiceBean;
+import edu.harvard.iq.dataverse.actionlogging.ActionLogServiceBean;
+import edu.harvard.iq.dataverse.engine.command.DataverseRequest;
+import edu.harvard.iq.dataverse.engine.command.exception.CommandException;
+import edu.harvard.iq.dataverse.engine.command.exception.PermissionException;
+import edu.harvard.iq.dataverse.persistence.ActionLogRecord;
+import edu.harvard.iq.dataverse.persistence.ActionLogRecord.ActionType;
+import edu.harvard.iq.dataverse.persistence.ActionLogRecord.Result;
+
+import javax.annotation.Resource;
+import javax.ejb.EJBContext;
+import javax.inject.Inject;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+import javax.validation.ConstraintViolationException;
+import java.lang.reflect.Method;
+import java.util.Date;
+import java.util.function.Supplier;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static edu.harvard.iq.dataverse.interceptors.InterceptorCommons.createName;
+
+@Interceptor
+@LoggedCall
+public class LoggedCallInterceptor {
+    private static final Logger logger = Logger.getLogger(LoggedCallInterceptor.class.getSimpleName());
+
+    @Inject
+    private ActionLogServiceBean logService;
+
+    @Inject
+    private DataverseRequestServiceBean requestService;
+
+    @Resource
+    EJBContext ejbContext;
+
+    // -------------------- LOGIC --------------------
+
+    @AroundInvoke
+    public Object callRestricted(InvocationContext ctx) throws Exception {
+        Method method = ctx.getMethod();
+        DataverseRequest request = requestService.getDataverseRequest();
+        ActionLogRecord logRecord = new ActionLogRecord(ActionType.Command, createName(method))
+                .setUserIdentifier(request.getUser().getIdentifier());
+        ctx.getContextData().put(InterceptorCommons.LOG_RECORD_KEY, logRecord);
+        try {
+            return ctx.proceed();
+        } catch(PermissionException pe) {
+            logRecord.setActionResult(Result.PermissionError);
+            addExceptionMessage(logRecord, pe::getMessage);
+            throw pe;
+        } catch (CommandException ce) {
+            logRecord.setActionResult(Result.InternalError);
+            addExceptionMessage(logRecord, ce::getMessage);
+            throw ce;
+        } catch (RuntimeException re) {
+            logRecord.setActionResult(Result.InternalError);
+            addExceptionMessage(logRecord, re::getMessage);
+            logConstraintViolations(logRecord, re);
+            throw re;
+        } finally {
+            if (logRecord.getActionResult() == null) {
+                logRecord.setActionResult(Result.OK);
+            } else {
+                ejbContext.setRollbackOnly();
```

Comment:
Not sure if setting .setRollbackOnly is needed does ejb won't rollback when some exception will occur?

Reply:
I've checked, and the answer is that theoretically it should work that way with some exceptions in case of caught exceptions – however because of possible corner cases I'd prefer to leave it as it is.

Reply:
Ok.